executor/slow_query.go

```diff
@@ -186,35 +181,119 @@ func (sc *slowLogChecker) isTimeValid(t types.Time) bool {
 	return true
 }
 
-// TODO: optimize for parse huge log-file.
-func (e *slowQueryRetriever) parseSlowLog(ctx sessionctx.Context, reader *bufio.Reader, maxRow int) ([][]types.Datum, error) {
-	var rows [][]types.Datum
-	var st *slowQueryTuple
-	startFlag := false
-	tz := ctx.GetSessionVars().Location()
-	for {
-		if len(rows) >= maxRow {
-			return rows, nil
+func getOneLine(reader *bufio.Reader) ([]byte, error) {
+	var resByte []byte
+	lineByte, isPrefix, err := reader.ReadLine()
+	if isPrefix {
+		// Need to read more data.
+		resByte = make([]byte, len(lineByte), len(lineByte)*2)
+	} else {
+		resByte = make([]byte, len(lineByte))
+	}
+	// Use copy here to avoid shallow copy problem.
+	copy(resByte, lineByte)
+	if err != nil {
+		return resByte, err
+	}
+
+	var tempLine []byte
+	for isPrefix {
+		tempLine, isPrefix, err = reader.ReadLine()
+		resByte = append(resByte, tempLine...)
+		// Use the max value of max_allowed_packet to check the single line length.
+		if len(resByte) > int(variable.MaxOfMaxAllowedPacket) {
+			return resByte, errors.Errorf("single line length exceeds limit: %v", variable.MaxOfMaxAllowedPacket)
 		}
-		e.fileLine++
-		lineByte, err := getOneLine(reader)
 		if err != nil {
-			if err == io.EOF {
-				e.fileIdx++
-				e.fileLine = 0
-				if e.fileIdx >= len(e.files) {
-					return rows, nil
+			return resByte, err
+		}
+	}
+	return resByte, err
+}
+
+func (e *slowQueryRetriever) getBatchLog(reader *bufio.Reader, num int) ([]string, error) {
+	var line string
+	log := make([]string, 0, num)
+	var err error
+	for i := 0; i < num; i++ {
+		for {
+			e.fileLine++
+			lineByte, err := getOneLine(reader)
+			if err != nil {
+				if err == io.EOF {
+					e.fileIdx++
+					e.fileLine = 0
+					if e.fileIdx >= len(e.files) {
+						return log, nil
+					}
+					reader.Reset(e.files[e.fileIdx].file)
+					continue
 				}
-				reader.Reset(e.files[e.fileIdx].file)
-				continue
+				return log, err
+			}
+			line = string(hack.String(lineByte))
+			log = append(log, line)
+			if strings.HasSuffix(line, variable.SlowLogSQLSuffixStr) {
+				if strings.HasPrefix(line, "use") {
+					continue
+				}
+				break
 			}
-			return rows, err
 		}
-		line := string(hack.String(lineByte))
-		// Check slow log entry start flag.
+	}
+	return log, err
+}
+
+func (e *slowQueryRetriever) parseSlowLog(ctx context.Context, sctx sessionctx.Context, reader *bufio.Reader, logNum int) {
+	// To limit the num of go routine
+	var wg sync.WaitGroup
+	offset := 0
+	ch := make(chan int, sctx.GetSessionVars().Concurrency.DistSQLScanConcurrency())
+	defer close(ch)
+	for {
+		log, err := e.getBatchLog(reader, logNum)
+		wg.Add(1)
+		ch <- 1
+		if err != nil {
+			go func() {
+				defer wg.Done()
+				e.parsedSlowLogCh <- parsedSlowLog{nil, err}
+				<-ch
+			}()
+			break
+		}
+		start := offset
+		go func() {
+			defer wg.Done()
+			e.parsedSlowLogCh <- parsedSlowLog{e.parsedLog(sctx, log, start), err}
```

Comment:
What happened if `parsedSlowLog` panic?

Reply:
> What happened if `parsedSlowLog` panic?

when parsedLog setFieldValue error, it will append warning to StatementContext and continue parse log just like this:
```
valid, err := st.setFieldValue(tz, variable.SlowLogUserAndHostStr, value, offset+index+1, e.checker)
if err != nil {
       ctx.GetSessionVars().StmtCtx.AppendWarning(err)
       continue
}
```

Reply:
I mean `panic` in function `parsedSlowLog`, not just a parse error.